In [ ]:
import pandas as pd
BigVuln = pd.read_csv("Big Vuln/MSR_data_cleaned.csv") # full BigVuln data
MegaVuln = pd.read_json("Mega Vuln/megavul.json") # full MegaVuln data

## Basic Analysis 

### BigVuln

In [2]:
BigVuln.shape , BigVuln.columns

((188636, 36),
 Index(['Unnamed: 0', 'Access Gained', 'Attack Origin',
        'Authentication Required', 'Availability', 'CVE ID', 'CVE Page',
        'CWE ID', 'Complexity', 'Confidentiality', 'Integrity',
        'Known Exploits', 'Publish Date', 'Score', 'Summary', 'Update Date',
        'Vulnerability Classification', 'add_lines', 'codeLink', 'commit_id',
        'commit_message', 'del_lines', 'file_name', 'files_changed',
        'func_after', 'func_before', 'lang', 'lines_after', 'lines_before',
        'parentID', 'patch', 'project', 'project_after', 'project_before',
        'vul', 'vul_func_with_fix'],
       dtype='object'))

In [3]:
BigVuln.head()

,Unnamed: 0,Access Gained,Attack Origin,Authentication Required,Availability,CVE ID,CVE Page,CWE ID,Complexity,Confidentiality,...,lang,lines_after,lines_before,parentID,patch,project,project_after,project_before,vul,vul_func_with_fix
0,0,None,Remote,Single system,Partial,CVE-2015-8467,https://www.cvedetails.com/cve/CVE-2015-8467/,CWE-264,Medium,Partial,...,C,NaN,NaN,a819d2b440aafa3138d95ff6e8b824da885a70e9,"@@ -1558,12 +1558,15 @@ static int samldb_chec...",samba,https://git.samba.org/?p=samba.git;a=blob;f=so...,https://git.samba.org/?p=samba.git;a=blob;f=so...,0,static bool check_rodc_critical_attribute(stru...
1,1,None,Remote,Single system,Partial,CVE-2015-8467,https://www.cvedetails.com/cve/CVE-2015-8467/,CWE-264,Medium,Partial,...,C,NaN,NaN,a819d2b440aafa3138d95ff6e8b824da885a70e9,"@@ -1558,12 +1558,15 @@ static int samldb_chec...",samba,https://git.samba.org/?p=samba.git;a=blob;f=so...,https://git.samba.org/?p=samba.git;a=blob;f=so...,0,static int samldb_add_entry(struct samldb_ctx ...
2,2,None,Remote,Single system,Partial,CVE-2015-8467,https://www.cvedetails.com/cve/CVE-2015-8467/,CWE-264,Medium,Partial,...,C,NaN,NaN,a819d2b440aafa3138d95ff6e8b824da885a70e9,"@@ -1558,12 +1558,15 @@ static int samldb_chec...",samba,https://git.samba.org/?p=samba.git;a=blob;f=so...,https://git.samba.org/?p=samba.git;a=blob;f=so...,0,static int samldb_add_entry_callback(struct ld...
3,3,None,Remote,Single system,Partial,CVE-2015-8467,https://www.cvedetails.com/cve/CVE-2015-8467/,CWE-264,Medium,Partial,...,C,NaN,NaN,a819d2b440aafa3138d95ff6e8b824da885a70e9,"@@ -1558,12 +1558,15 @@ static int samldb_chec...",samba,https://git.samba.org/?p=samba.git;a=blob;f=so...,https://git.samba.org/?p=samba.git;a=blob;f=so...,0,static int samldb_add_handle_msDS_IntId(struct...
4,4,None,Remote,Single system,Partial,CVE-2015-8467,https://www.cvedetails.com/cve/CVE-2015-8467/,CWE-264,Medium,Partial,...,C,NaN,NaN,a819d2b440aafa3138d95ff6e8b824da885a70e9,"@@ -1558,12 +1558,15 @@ static int samldb_chec...",samba,https://git.samba.org/?p=samba.git;a=blob;f=so...,https://git.samba.org/?p=samba.git;a=blob;f=so...,0,static int samldb_add_step(struct samldb_ctx *...


#### summarize each column

In [11]:
def summarize_dataset(df):
    summary = []
    for column in df.columns:
        num_unique_values = df[column].nunique()
        
        col_data = {}
        col_data['Column Name'] = column
        col_data['Data Type'] = str(df[column].dtype)
        col_data['Total Number of Unique Values'] = num_unique_values
        
        # Skip writing Unique Values if there are more than 200 unique values
        if num_unique_values <= 200:
            col_data['Unique Values'] = df[column].unique().tolist()
        else:
            col_data['Unique Values'] = "Too many unique values to display"
        
        if df[column].dtype in ['int64', 'float64']:
            col_data['Statistics'] = {
                'Mean': df[column].mean(),
                'Min': df[column].min(),
                'Max': df[column].max(),
            }
        summary.append(col_data)

    # Convert the summary to a formatted string in Markdown format
    summary_text = ""
    for col in summary:
        summary_text += f"## {col['Column Name']}\n"
        summary_text += f"### Data Type: {col['Data Type']}\n"
        summary_text += f"### Total Number of Unique Values: {col['Total Number of Unique Values']}\n"
        summary_text += f"### Unique Values: {col['Unique Values']}\n"
        if 'Statistics' in col:
            stats = col['Statistics']
            summary_text += "### Statistics:\n"
            summary_text += f"#### Mean: {stats['Mean']}\n"
            summary_text += f"#### Min: {stats['Min']}\n"
            summary_text += f"#### Max: {stats['Max']}\n"
        summary_text += "\n"

    # Save the summary to a .md file with utf-8 encoding
    with open('BigVuln_dataset_summary.md', 'w', encoding='utf-8') as file:
        file.write(summary_text)

# Assuming you have your DataFrame named 'BigVuln'
summarize_dataset(BigVuln)

### MegaVuln

In [4]:
MegaVuln.shape , MegaVuln.columns

((353873, 32),
 Index(['cve_id', 'cwe_ids', 'cvss_vector', 'cvss_base_score',
        'cvss_base_severity', 'cvss_is_v3', 'publish_date', 'repo_name',
        'commit_msg', 'commit_hash', 'parent_commit_hash', 'commit_date',
        'git_url', 'file_path', 'func_name', 'parameter_list_signature_before',
        'parameter_list_before', 'return_type_before', 'func_before',
        'abstract_func_before', 'abstract_symbol_table_before',
        'func_graph_path_before', 'parameter_list_signature', 'parameter_list',
        'return_type', 'func', 'abstract_func', 'abstract_symbol_table',
        'func_graph_path', 'diff_func', 'diff_line_info', 'is_vul'],
       dtype='object'))

In [5]:
MegaVuln.head()

,cve_id,cwe_ids,cvss_vector,cvss_base_score,cvss_base_severity,cvss_is_v3,publish_date,repo_name,commit_msg,commit_hash,...,parameter_list_signature,parameter_list,return_type,func,abstract_func,abstract_symbol_table,func_graph_path,diff_func,diff_line_info,is_vul
0,CVE-2016-2538,[CWE-189],AV:L/AC:L/Au:N/C:P/I:N/A:P,3.6,LOW,0.0,2016-06-16T18:59:06.093,qemu,usb: check RNDIS buffer offsets & length\n\nWh...,fe3c546c5ff2a6210f9a4d8561cc64051ca8603e,...,"(USBNetState *s, rndis_query_msg_type *buf, un...","[[USBNetState *, s, 13], [rndis_query_msg_type...",int,static int rndis_query_response(USBNetState *s...,static int rndis_query_response(USBNetState *V...,"{'position_map': {'0': [[7, -1, 'TYPE', 'int',...",qemu/fe3c546c5ff2a6210f9a4d8561cc64051ca8603e/...,"--- func_before\n+++ func_after\n@@ -10,8 +10,...",{'deleted_lines': [' if (bufoffs + buflen >...,True
1,CVE-2016-2538,[CWE-189],AV:L/AC:L/Au:N/C:P/I:N/A:P,3.6,LOW,0.0,2016-06-16T18:59:06.093,qemu,usb: check RNDIS buffer offsets & length\n\nWh...,fe3c546c5ff2a6210f9a4d8561cc64051ca8603e,...,"(USBNetState *s, rndis_set_msg_type *buf, unsi...","[[USBNetState *, s, 13], [rndis_set_msg_type *...",int,"static int rndis_set_response(USBNetState *s,\...",static int rndis_set_response(USBNetState *VAR...,"{'position_map': {'0': [[7, -1, 'TYPE', 'int',...",qemu/fe3c546c5ff2a6210f9a4d8561cc64051ca8603e/...,"--- func_before\n+++ func_after\n@@ -11,8 +11,...",{'deleted_lines': [' if (bufoffs + buflen >...,True
2,CVE-2016-2538,[CWE-189],AV:L/AC:L/Au:N/C:P/I:N/A:P,3.6,LOW,0.0,2016-06-16T18:59:06.093,qemu,usb: check RNDIS buffer offsets & length\n\nWh...,fe3c546c5ff2a6210f9a4d8561cc64051ca8603e,...,"(USBNetState *s, USBPacket *p)","[[USBNetState *, s, 13], [USBPacket *, p, 11]]",void,static void usb_net_handle_dataout(USBNetState...,static void usb_net_handle_dataout(USBNetState...,"{'position_map': {'0': [[7, -1, 'TYPE', 'void'...",qemu/fe3c546c5ff2a6210f9a4d8561cc64051ca8603e/...,"--- func_before\n+++ func_after\n@@ -30,8 +30,...",{'deleted_lines': [' if (offs + size <=...,True
3,CVE-2016-2841,[CWE-20],AV:L/AC:L/Au:N/C:N/I:N/A:P,2.1,LOW,0.0,2016-06-16T18:59:07.203,qemu,net: ne2000: check ring buffer control registe...,415ab35a441eca767d033a2702223e785b9d5190,...,(NE2000State *s),"[[NE2000State *, s, 13]]",int,static int ne2000_buffer_full(NE2000State *s)\...,static int ne2000_buffer_full(NE2000State *VAR...,"{'position_map': {'0': [[7, -1, 'TYPE', 'int',...",qemu/415ab35a441eca767d033a2702223e785b9d5190/...,"--- func_before\n+++ func_after\n@@ -1,6 +1,10...","{'deleted_lines': [], 'added_lines': ['', ' ...",True
4,CVE-2016-2841,[CWE-20],AV:L/AC:L/Au:N/C:N/I:N/A:P,2.1,LOW,0.0,2016-06-16T18:59:07.203,qemu,net: ne2000: check ring buffer control registe...,415ab35a441eca767d033a2702223e785b9d5190,...,"(NE2000State *s, uint32_t addr)","[[NE2000State *, s, 13], [uint32_t, addr, 9]]",uint32_t,static inline uint32_t ne2000_mem_readl(NE2000...,static inline uint32_t ne2000_mem_readl(NE2000...,"{'position_map': {'0': [[14, -1, 'TYPE', 'uint...",qemu/415ab35a441eca767d033a2702223e785b9d5190/...,None,None,False


#### Summarize each column

In [14]:
def summarize_dataset(df):
    summary = []
    for column in df.columns:
        try:
            num_unique_values = df[column].nunique()
        except TypeError:
            num_unique_values = 'N/A (contains unhashable types)'
        
        col_data = {}
        col_data['Column Name'] = column
        col_data['Data Type'] = str(df[column].dtype)
        col_data['Total Number of Unique Values'] = num_unique_values
        
        if isinstance(num_unique_values, int) and num_unique_values <= 200:
            col_data['Unique Values'] = df[column].unique().tolist()
        else:
            col_data['Unique Values'] = "Too many unique values to display" if isinstance(num_unique_values, int) else num_unique_values
        
        if df[column].dtype in ['int64', 'float64']:
            col_data['Statistics'] = {
                'Mean': df[column].mean(),
                'Min': df[column].min(),
                'Max': df[column].max(),
                'Std Dev': df[column].std()
            }
        summary.append(col_data)

    summary_text = ""
    for col in summary:
        summary_text += f"## {col['Column Name']}\n"
        summary_text += f"### Data Type: {col['Data Type']}\n"
        summary_text += f"### Total Number of Unique Values: {col['Total Number of Unique Values']}\n"
        summary_text += f"### Unique Values: {col['Unique Values']}\n"
        if 'Statistics' in col:
            stats = col['Statistics']
            summary_text += "### Statistics:\n"
            summary_text += f"#### Mean: {stats['Mean']}\n"
            summary_text += f"#### Min: {stats['Min']}\n"
            summary_text += f"#### Max: {stats['Max']}\n"
            summary_text += f"#### Std Dev: {stats['Std Dev']}\n"
        summary_text += "\n"

    with open('MegaVuln_dataset_summary.md', 'w', encoding='utf-8') as file:
        file.write(summary_text)

summarize_dataset(MegaVuln)


## Find data intersction  

In [15]:
BigVuln['commit_id'].loc[0]

'b000da128b5fb519d2d3f2e7fd20e4a25b7dae7d'

In [17]:
MegaVuln['commit_hash'].loc[0]

'fe3c546c5ff2a6210f9a4d8561cc64051ca8603e'

### General intersection

In [39]:
hashes_megavuln = MegaVuln[['commit_hash']]
hashes_bigvuln = BigVuln[['commit_id']]
print("Total Number of hashes in MegaVuln :", len(hashes_megavuln)) 
print("Total Number of hashes in BigVuln :", len(hashes_bigvuln))

unique_megavuln = MegaVuln[['commit_hash']].drop_duplicates()
unique_bigvuln = BigVuln[['commit_id']].drop_duplicates()
print("Total Unique hashs in MegaVuln :",len(unique_megavuln))
print("Total Unique hashs in BigVuln :",len(unique_bigvuln))

intersected_rows = pd.merge(
    unique_megavuln,
    unique_bigvuln,
    left_on='commit_hash',
    right_on='commit_id'
)

num_intersected_rows = len(intersected_rows)

print(f"Number of intersected rows based on commit hash: {num_intersected_rows}")

Total Number of hashes in MegaVuln : 353873
Total Number of hashes in BigVuln : 188636
Total Unique hashs in MegaVuln : 9288
Total Unique hashs in BigVuln : 4058
Number of intersected rows based on commit hash: 2513


### Vulnurable functions intersection

In [50]:
vulnerable_megavuln = MegaVuln[MegaVuln['is_vul'] == True]
vulnerable_bigvuln = BigVuln[BigVuln['vul'] == 1]
print("Total Number of Vulnurable functions in MegaVuln :", len(vulnerable_megavuln)) 
print("Total Number of Vulnurable functions in BigVuln :", len(vulnerable_bigvuln))

vulnerable_megavuln = vulnerable_megavuln[['commit_hash']].drop_duplicates()
vulnerable_bigvuln = vulnerable_bigvuln[['commit_id']].drop_duplicates()
print("Total Unique hashs in MegaVuln :",len(vulnerable_megavuln))
print("Total Unique hashs in BigVuln :",len(vulnerable_bigvuln))

intersectionVuln = pd.merge(
    vulnerable_bigvuln,
    vulnerable_megavuln,
    left_on='commit_id',
    right_on='commit_hash'
)

num_intersected_rows = len(intersectionVuln)

print(f"Number of intersected rows based on commit hash: {num_intersected_rows}")

Total Number of Vulnurable functions in MegaVuln : 17975
Total Number of Vulnurable functions in BigVuln : 10900
Total Unique hashs in MegaVuln : 9288
Total Unique hashs in BigVuln : 3746
Number of intersected rows based on commit hash: 2391


In [2]:
vulnerable_megavuln = MegaVuln[MegaVuln['is_vul'] == True]
vulnerable_bigvuln = BigVuln[BigVuln['vul'] == 1]

print("Total Number of Vulnerable functions in MegaVuln:", len(vulnerable_megavuln))
print("Total Number of Vulnerable functions in BigVuln:", len(vulnerable_bigvuln))

megavuln_counts = vulnerable_megavuln['commit_hash'].value_counts()

bigvuln_counts = vulnerable_bigvuln['commit_id'].value_counts()

num_intersected_rows = 0

# Iterate over each commit hash in megavuln_counts and check if it exists in bigvuln_counts
for commit_hash in megavuln_counts.index:
    if commit_hash in bigvuln_counts:
        num_intersected_rows += min(megavuln_counts[commit_hash], bigvuln_counts[commit_hash])

print(f"Number of intersected rows vulnerable functions only (func_before) based on commit hash: {num_intersected_rows}")


Total Number of Vulnerable functions in MegaVuln: 17975
Total Number of Vulnerable functions in BigVuln: 10900
Number of intersected rows vulnerable functions only (func_before) based on commit hash: 4190


In [17]:
print("max Value of commit hash occurrences in MegaVuln",megavuln_counts.max())
print("Average  Value of commit hash occurrences in MegaVuln",megavuln_counts.mean())
print("min Value of commit hash occurrences in MegaVuln",megavuln_counts.min())
print("max Value of commit hash occurrences in BigVuln",bigvuln_counts.max())
print("Average  Value of commit hash occurrences in BigVuln",bigvuln_counts.mean())
print("min Value of commit hash occurrences in BigVuln",bigvuln_counts.min())

max Value of commit hash occurrences in MegaVuln 33
Average  Value of commit hash occurrences in MegaVuln 1.9352928509905254
min Value of commit hash occurrences in MegaVuln 1
max Value of commit hash occurrences in BigVuln 237
Average  Value of commit hash occurrences in BigVuln 2.9097704217832354
min Value of commit hash occurrences in BigVuln 1


In [8]:
# Get intersected commit hashes
intersected_commit_hashes = megavuln_counts.index.intersection(bigvuln_counts.index)

# Filter datasets for these commit hashes
intersected_megavuln = vulnerable_megavuln[vulnerable_megavuln['commit_hash'].isin(intersected_commit_hashes)]
intersected_bigvuln = vulnerable_bigvuln[vulnerable_bigvuln['commit_id'].isin(intersected_commit_hashes)]
print(len(intersected_megavuln))
print(len(intersected_bigvuln))

4471
4424


In [14]:
vulnerable_megavuln = MegaVuln[MegaVuln['is_vul'] == True]
vulnerable_bigvuln = BigVuln[BigVuln['vul'] == 1]

print("Total Number of Vulnerable functions in MegaVuln:", len(vulnerable_megavuln))
print("Total Number of Vulnerable functions in BigVuln:", len(vulnerable_bigvuln))

# Count commit hash occurrences
megavuln_counts = vulnerable_megavuln['commit_hash'].value_counts()
bigvuln_counts = vulnerable_bigvuln['commit_id'].value_counts()

# Find intersecting commit hashes
intersected_commits = list(set(megavuln_counts.index).intersection(set(bigvuln_counts.index)))
print(f"Intersected Commits: {len(intersected_commits)}")

megavuln_indices = []
bigvuln_indices = []
num_intersected_rows = 0

for commit_hash in intersected_commits:
    # Get all indices (with duplicates) of the intersecting commit hash from both datasets
    mega_indices = vulnerable_megavuln[vulnerable_megavuln['commit_hash'] == commit_hash].index.tolist()
    big_indices = vulnerable_bigvuln[vulnerable_bigvuln['commit_id'] == commit_hash].index.tolist()
    
    # Store indices
    megavuln_indices.extend(mega_indices)
    bigvuln_indices.extend(big_indices)
    
    # Calculate the intersection count (min for each commit)
    num_intersected_rows += min(len(mega_indices), len(big_indices))

# Print results
print(f"Number of intersected rows vulnerable functions only (func_before) based on commit hash: {num_intersected_rows}")
print(f"Total intersected indices in MegaVuln: {len(megavuln_indices)}")
print(f"Total intersected indices in BigVuln: {len(bigvuln_indices)}")

Total Number of Vulnerable functions in MegaVuln: 17975
Total Number of Vulnerable functions in BigVuln: 10900
Intersected Commits: 2391
Number of intersected rows vulnerable functions only (func_before) based on commit hash: 4190
Total intersected indices in MegaVuln: 4471
Total intersected indices in BigVuln: 4424


In [15]:
# Save the data to a file 
with open('intersected indices in MegaVuln.txt', 'w') as file:
    for index in megavuln_indices:
        file.write(f"{index}\n")

with open('intersected indices in BigVuln.txt', 'w') as file:
    for index in bigvuln_indices:
        file.write(f"{index}\n")

In [41]:
set(intersectionVuln).issubset(set(intersected_rows))

True

### CWE analysis

In [61]:
MegaVuln_cwe_ids = [189, 20, 119, 400, 704, 190,
    787, 476, 125, 416, 126, 122, 415,
    617, 401, 330, 755, 770, 908, 121,
    369, 843, 200, 212, 129, 835, 131,
    667, 193, 362, 252, 674, 388, 276,
    754, 502, 732, 59, 78, 185, 113,
    116, 94, 668, 611, 665, 601, 79,
    326, 269, 347, 22, 191, 77, 1049,
    776, 327, 203, 862, 662, 682, 909,
    628, 120, 345, 918, 459, 772, 863,
    134, 264, 287, 1284, 823, 763, 552,
    693, 697, 284, 202, 273, 367, 248,
    681, 1188, 404, 427, 295, 319, 399,
    310, 16, 494, 254, 19, 834, 532,
    426, 1077, 74, 255, 89, 672, 626,
    440, 762, 1333, 428, 17, 361, 331,
    824, 1021, 436, 118, 670, 444, 434,
    306, 680, 338, 460, 290, 208, 337,
    73, 522, 90, 924, 324, 613, 911,
    241, 475, 285, 172, 320, 352, 93,
    358, 346, 1187, 335, 117, 385, 88,
    209, 184, 590, 639, 312, 788, 409,
    1050, 282, 323, 610, 354, 18, 943,
    407, 250, 384, 300, 26, 23, 229,
    197, 307, 913, 325, 838, 706, 349,
    707]


BigVuln_cwe_ids = [264, 119, 310, 189, 200,
        20, 416, 399, 835, 125, 347,
        190, 94, 17, 362, 704, 354,
        78, 59, 269, 388, 617, 476,
        769, 787, 754, 502, 22, 665,
        254, 772, 290, 400, 369, 284,
        93, 415, 674, 311, 330, 327,
        320, 77, 134, 732, 287, 16,
        89, 255, 18, 19, 285, 74,
        79, 404, 862, 834, 494, 755,
        129, 191, 682, 522, 918, 358,
        295, 770, 346, 532, 426, 824,
        693, 352, 436, 90, 763, 120,
        611, 601, 706, 361, 281, 1021,
        668, 664, 172, 909, 209, 345,
        331, 252]

mitre_c_cwe_ids = [
    14, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 134, 135, 170, 188, 191, 192, 
    194, 195, 196, 197, 242, 243, 244, 362, 364, 366, 374, 375, 401, 415, 416, 457, 460, 462, 463, 464, 
    466, 467, 468, 469, 474, 476, 478, 479, 480, 481, 482, 483, 484, 495, 496, 558, 560, 562, 587, 676, 
    685, 688, 689, 690, 704, 733, 762, 781, 782, 783, 785, 787, 789, 805, 806, 839, 843, 910, 911, 1325, 
    1335, 1341
]

mitre_cpp_cwe_ids = [
    14, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 134, 135, 170, 188, 191, 192, 
    194, 195, 196, 197, 242, 243, 244, 248, 362, 364, 366, 374, 375, 396, 397, 401, 415, 416, 457, 460, 
    462, 463, 464, 466, 467, 468, 469, 476, 478, 479, 480, 481, 482, 483, 484, 493, 495, 496, 498, 500, 
    543, 558, 562, 587, 676, 690, 704, 733, 762, 766, 767, 781, 782, 783, 785, 787, 789, 805, 806, 839, 
    843, 910, 911, 1325, 1335, 1341
]

mitre_all = list(set(mitre_c_cwe_ids + mitre_cpp_cwe_ids))
intersectionAll = list(set(mitre_all) & set(BigVuln_cwe_ids) & set(MegaVuln_cwe_ids))
intersection = list(set(BigVuln_cwe_ids) & set(MegaVuln_cwe_ids))
isSubset = set(BigVuln_cwe_ids).issubset(set(MegaVuln_cwe_ids))
print("Total number of intersected CWE in BigVuln:",len(BigVuln_cwe_ids))
print("Total number of intersected CWE in MegaVuln:",len(MegaVuln_cwe_ids))
print("Total number of intersected CWE between Mitre & BigVuln & MegaVuln:",len(intersectionAll))
print("List of intersected CWE between Mitre & BigVuln & MegaVuln:",intersectionAll)
print("Total number of intersected CWE between BigVuln & MegaVuln:",len(intersection))
print("List of intersected CWE between BigVuln & MegaVuln:\n",intersection)
print("is BigVuln a subset of MegaVuln:",isSubset) 

Total number of intersected CWE in BigVuln: 91
Total number of intersected CWE in MegaVuln: 175
Total number of intersected CWE between Mitre & BigVuln & MegaVuln: 12
List of intersected CWE between Mitre & BigVuln & MegaVuln: [704, 129, 416, 191, 134, 362, 787, 119, 120, 476, 125, 415]
Total number of intersected CWE between BigVuln & MegaVuln: 87
List of intersected CWE between BigVuln & MegaVuln:
 [522, 16, 17, 18, 19, 20, 532, 22, 59, 74, 77, 78, 79, 601, 89, 90, 93, 94, 611, 617, 119, 120, 125, 129, 134, 665, 668, 674, 682, 172, 693, 189, 190, 191, 704, 706, 200, 209, 732, 754, 755, 763, 252, 254, 255, 770, 772, 264, 269, 787, 284, 285, 287, 290, 295, 310, 824, 320, 834, 835, 327, 330, 331, 345, 346, 347, 862, 352, 354, 358, 361, 362, 369, 388, 909, 399, 400, 404, 918, 415, 416, 426, 436, 476, 494, 502, 1021]
is BigVuln a subset of MegaVuln: False


Duplicates in Raw Data: The raw counts (**353,873** in `MegaVuln` vs. **9,288 unique values**, and **188,636** in `BigVuln` vs. **4,058 unique values**) show that duplicates dominate the data. These duplicates artificially inflate the dataset size but do not affect the unique intersection count once deduplication is applied.

Overlap Between Unique Hashes:
After removing duplicates, `MegaVuln` has **9,288** unique hashes, and `BigVuln` has **4,058**.
The intersection count of **2,513** suggests that not all hashes are shared between the datasets. In fact:
2,513 / 4,058 ≈ **62%** of `BigVuln`'s unique hashes are found in `MegaVuln`.
2,513 / 9,288 ≈ **27%** of `MegaVuln`'s unique hashes are found in `BigVuln`.
2,513 / 2,391 ≈ **955** of intersected unique hashes are **Vulnurable**.

|Features                     | MegaVul(C/CPP) | Big-Vul        |
|-----------------------------|----------------|----------------|
| Number of Repositories      | 1062           | 310            |
| Number of CVE IDs           | 8476           | 3539           |
| Number of CWE IDs           | 176            | 92             |
| Number of Commits           | 9288           | 4058           |
| Number of Vul/Non-Vul Function | 17975/335898   | 10900/177736   |
| Success Rate of Graph Generation | 87%            | None           |
